### PR TITLE
[IMP] add a keychain backend abstract model

### PR DESCRIPTION
**Pull request by mourad-ehm**
the aim is to simplify the integration of keychain when you already have /need the concept of backend.

### CODE REVIEW

File: keychain/models/keychain.py [hunk 1]
```diff
@@ -200,3 +200,71 @@ def _get_keys(envs):
                 "Use a key similar to: %s" % (envs[0], Fernet.generate_key())
             ))
         return MultiFernet(keys)
+
+
+class KeychainBackend(models.AbstractModel):
```
Comment (hparfr) [inline keychain/models/keychain.py]: put in a separate file

File: keychain/models/keychain.py [hunk 2]
```diff
@@ -200,3 +200,71 @@ def _get_keys(envs):
...
+        for record in self:
+            account = self._get_keychain_account()
+            if record.password and record.password != '******':
+                account.clear_password = record.password
+
+    def _compute_password(self):
+        for record in self:
+            account = record._get_existing_keychain()
+            if account and account.password:
+                record.password = "******"
```
Comment (hparfr) [inline keychain/models/keychain.py]: from a ui point of view, I don't like these **  :  Some user will think theirs passwords is 6 char long.

File: keychain/tests/test_keychain.py [hunk 3]
```diff
@@ -218,3 +230,40 @@ def test_valid_json(self):
                 self.assertTrue(True, 'Should validate json')
             except:
                 self.assertTrue(False, 'It should validate a good json')
+
+    def test_keychain_bakend(self):
```
Comment (hparfr) [inline keychain/tests/test_keychain.py]: you may  put in a separate file

File: keychain/models/keychain_backend.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    name = fields.Char(required=True)
+    password = fields.Char(
+        compute="_compute_password",
+        inverse="_inverse_password",
+        required=True)
+    data = fields.Serialized(
+        compute="_compute_keychain",
+        inverse="_inverse_keychain")
+
+    def _get_technical_name(self):
```
Comment (bealdav) [inline keychain/models/keychain_backend.py]: missing api.model

File: keychain/models/keychain_backend.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        return '%s,%s' % (self._name, self.id)
+
+    def _get_existing_keychain(self):
+        self.ensure_one()
+        return self.env['keychain.account'].retrieve([
+            ('namespace', '=', self._backend_name),
+            ('technical_name', '=', self._get_technical_name())
+        ])
+
+    def _prepare_keychain(self):
```
Comment (bealdav) [inline keychain/models/keychain_backend.py]: missing api.model

File: keychain/models/keychain_backend.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    _backend_name = None
+
+    name = fields.Char(required=True)
+    password = fields.Char(
+        compute="_compute_password",
+        inverse="_inverse_password",
+        required=True)
+    data = fields.Serialized(
+        compute="_compute_keychain",
+        inverse="_inverse_keychain")
```
Comment (bealdav) [inline keychain/models/keychain_backend.py]: require help argument

File: keychain/tests/__init__.py [hunk 7]
```diff
@@ -1 +1,2 @@
-from . import test_keychain
\ No newline at end of file
+from . import test_keychain
+from . import test_keychain_backend
```
Comment (bealdav) [inline keychain/tests/__init__.py]: missing line

**Review by bealdav — CHANGES_REQUESTED** [review]

File: keychain/models/keychain_backend.py (hunk 4 above)
Comment (hparfr) [reply keychain/models/keychain_backend.py]: no it's api.multi with a missing self.ensure_one()

File: keychain/models/keychain_backend.py (hunk 5 above)
Comment (hparfr) [reply keychain/models/keychain_backend.py]: missing self.ensure_one for an api.multi

**Review by hparfr — APPROVED** [review]

**Review by bealdav — APPROVED** [review]

**Pull request merged**